presto-hive/src/main/java/io/prestosql/plugin/hive/MsckProcedure.java

```diff
@@ -0,0 +1,213 @@
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static io.prestosql.plugin.hive.HiveMetadata.PRESTO_QUERY_ID_NAME;
+import static io.prestosql.plugin.hive.HivePartitionManager.extractPartitionValues;
+import static io.prestosql.plugin.hive.HiveSessionProperties.isRespectTableFormat;
+import static io.prestosql.plugin.hive.PartitionStatistics.empty;
+import static io.prestosql.plugin.hive.metastore.StorageFormat.fromHiveStorageFormat;
+import static io.prestosql.spi.block.MethodHandleUtil.methodHandle;
+import static io.prestosql.spi.type.StandardTypes.BOOLEAN;
+import static io.prestosql.spi.type.StandardTypes.VARCHAR;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class MsckProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle MSCK = methodHandle(
+            MsckProcedure.class,
+            "msck",
+            ConnectorSession.class,
+            String.class,
+            String.class,
+            Boolean.class,
+            Boolean.class);
+
+    private final Supplier<TransactionalMetadata> hiveMetadataFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+
+    @Inject
+    public MsckProcedure(
+            Supplier<TransactionalMetadata> hiveMetadataFactory,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "msck",
+                ImmutableList.of(
+                        new Argument("schema_name", VARCHAR),
+                        new Argument("table_name", VARCHAR),
+                        new Argument("add", BOOLEAN),
+                        new Argument("drop", BOOLEAN)),
+                MSCK.bindTo(this));
+    }
+
+    public void msck(ConnectorSession session, String schemaName, String tableName, Boolean add, Boolean drop)
+    {
+        HdfsEnvironment.HdfsContext context = new HdfsEnvironment.HdfsContext(session, schemaName, tableName);
+        SemiTransactionalHiveMetastore metastore = ((HiveMetadata) hiveMetadataFactory.get()).getMetastore();
+
+        Table table = metastore.getTable(schemaName, tableName).orElse(null);
+        if (table == null) {
+            return;
+        }
+
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation);
+
+            List<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName).orElse(emptyList());
```

Comment:
This can return Optional empty only when table does not exist, so 
`.orElseThrow(...)` would be more appropriate